execution/src/windows/execution.cc

```diff
@@ -221,17 +219,71 @@ namespace leatherman { namespace execution {
         const string name;
         HANDLE handle;
         OVERLAPPED overlapped;
-        scoped_resource<HANDLE> event;
+        scoped_handle event;
+        bool pending;
+        string buffer;
+        function<bool(string const&)> callback;
+    };
+
+    struct input_pipe
+    {
+        input_pipe(scoped_handle h, string b) :
+            handle(move(h)),
+            overlapped{},
+            pending(false),
+            buffer(move(b))
+        {
+            if (handle != INVALID_HANDLE_VALUE) {
+                event = scoped_handle(CreateEvent(nullptr, TRUE, FALSE, nullptr));
+                if (!event) {
+                    LOG_ERROR("failed to create stdin write event: %1%.", system_error());
+                    throw execution_exception("failed to create write event.");
+                }
+                overlapped.hEvent = event;
+            }
+        }
+
+        scoped_handle handle;
+        OVERLAPPED overlapped;
+        scoped_handle event;
         bool pending;
         string buffer;
-        function<bool(string const&)> const& callback;
     };
 
-    static void read_from_child(DWORD child, array<pipe, 2>& pipes, uint32_t timeout, HANDLE timer)
+    static void rw_from_child(DWORD child, array<pipe, 2>& pipes, input_pipe input, uint32_t timeout, HANDLE timer)
     {
         vector<HANDLE> wait_handles;
         while (true)
         {
+            // First try to write input
+            if (input.handle != INVALID_HANDLE_VALUE && !input.pending) {
+                while (true) {
+                    DWORD count;
+                    if (!WriteFile(input.handle, input.buffer.c_str(), input.buffer.size(), &count, &input.overlapped)) {
+                        // Treat broken pipes as closed pipes
+                        if (GetLastError() == ERROR_BROKEN_PIPE) {
+                            input.handle = {};
+                            break;
+                        }
+                        // Check to see if it's a pending operation
+                        if (GetLastError() == ERROR_IO_PENDING) {
+                            input.pending = true;
+                            break;
+                        }
+                        LOG_ERROR("failed to write child stdin input: %1%.", system_error());
+                        throw execution_exception("failed to write child process input.");
+                    }
+
+                    if (count == 0) {
+                        input.handle = {};
+                        break;
+                    }
+
+                    // Register written data
+                    input.buffer.erase(0, count);
+                }
+            }
+
```

Comment:
5 lines down is a bug, it should be continuing not breaking. That would explain much of the weird behavior I've been running into.